### PR TITLE
Fix OnTimeout event handling for CAN module

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
The problem is that `TimerDirector`'s timer checks for every request timeout every 10 seconds and not depends on time when request was sent. For example `TimeDirector` timer check all request for timeouts on 10, 20, 30 second. But if request was sent on 12th second, then default timeout for it will be on 22th second. However this timer will detect timeout and send response only on 30 second.
Also `OnTimeout` event provides wrong result code `TIMED_OUT` instead of `GENERIC_ERROR`.

In this pull request:
- Added more logs to see request controller actions
- Fixed result code for timeout event according to requirements
- Added `GetSecondsToNearestTimeout()` function to return time to the nearest request timeout or default timeout if no requests in queue. This is a correct time to wait before checking timeout. Also this function checks for custom RPC timeouts if present.
- Added `ResetTimer()` function to recalculate timers awaiting time after new request was added to queue

### CODE REVIEW

**Review by LuxoftAKutsan — APPROVED** [review]

File: src/components/can_cooperation/src/can_module.cc [hunk 1]
```diff
@@ -367,6 +367,7 @@ void CANModule::SendResponseToMobile(application_manager::MessagePtr msg) {
 
 void CANModule::SendTimeoutResponseToMobile(
     application_manager::MessagePtr msg) {
+  LOG4CXX_DEBUG(logger_, "Response to mobile: " << msg->json_message());
```
Comment (AByzhynar) [inline src/components/can_cooperation/src/can_module.cc]: @AKalinich-Luxoft How can we differentiate normal response to mobile and this one? Propose to change log message.

File: src/components/interfaces/HMI_API.xml [hunk 2]
```diff
@@ -63,7 +63,7 @@
     <element name="USER_DISALLOWED" value="23"/>
     <element name="TRUNCATED_DATA" value="24"/>
     <element name="SAVED" value="25"/>
-    <element name="READ_ONLY " value="26"/>
```
Comment (AByzhynar) [inline src/components/interfaces/HMI_API.xml]: @AKalinich-Luxoft How are this file and change related to timeouts?

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/can_cooperation/src/commands/base_command_request.cc [hunk 3]
```diff
@@ -75,15 +75,17 @@ void BaseCommandRequest::SendResponse(bool success,
   }
 
   msg_params[kSuccess] = success;
-  msg_params[kResultCode] = result_code;
+  const bool is_timeout = 0 == strcmp(result_code, result_codes::kTimedOut);
```
Comment (AByzhynar) [inline src/components/can_cooperation/src/commands/base_command_request.cc]: @AKalinich-Luxoft I suggest to add round brackets to improve readability

File: src/components/can_cooperation/src/request_controller.cc [hunk 4]
```diff
@@ -60,18 +60,26 @@ void RequestController::AddRequest(const uint32_t mobile_correlation_id,
...
+      logger_, "Delete request with correlation_id: " << mobile_correlation_id);
   mobile_request_list_.erase(mobile_correlation_id);
   // TODO(VS): add app id
   timer_.RemoveTrackable(TrackableMessage(0, mobile_correlation_id));
 }
 
 void RequestController::OnTimeoutTriggered(const TrackableMessage& expired) {
+  LOG4CXX_DEBUG(
+      logger_,
+      "Timeout for request with correlation_id: " << expired.correlation_id());
```
Comment (AByzhynar) [inline src/components/can_cooperation/src/request_controller.cc]: @AKalinich-Luxoft Timeout for request xxx - WHAT?

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/can_cooperation/src/can_module.cc (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/can_cooperation/src/can_module.cc]: @AByzhynar message updated in a522a1615baee23218f4a937b24d9459c680d605

File: src/components/interfaces/HMI_API.xml (hunk 2 above)
Comment (AKalinich-Luxoft) [reply src/components/interfaces/HMI_API.xml]: @AByzhynar this changes from another pull request. Rebased.

File: src/components/can_cooperation/src/commands/base_command_request.cc (hunk 3 above)
Comment (AKalinich-Luxoft) [reply src/components/can_cooperation/src/commands/base_command_request.cc]: @AByzhynar done in 6e49a4fed5df88d1b2f288205ec45ca9c131170d

File: src/components/can_cooperation/src/request_controller.cc (hunk 4 above)
Comment (AKalinich-Luxoft) [reply src/components/can_cooperation/src/request_controller.cc]: @AByzhynar fixed in a522a1615baee23218f4a937b24d9459c680d605

**Review by AByzhynar — APPROVED** [review]

**Pull request merged**